### ISSUE TITLE
Package stubber throwing exceptions

### ISSUE THREAD

**Issue by SamHatoum** [xolvio/jasmine-unit#16]
I just added velocity and jasmine-unit to our app and I first got:

```
I20140621-14:08:23.769(-7)? [PackageStubber] stubbing Layout
I20140621-14:08:23.789(-7)? Exception in queued task: Error: [PackageStubber.generateStubJsCode] Required field 'target' must be of type object or function
I20140621-14:08:23.789(-7)?     at Object._.extend.validate.generateStubJsCode (packages/package-stubber/main.js:88)
I20140621-14:08:23.790(-7)?     at Object._.extend.generateStubJsCode (packages/package-stubber/main.js:365)
I20140621-14:08:23.790(-7)?     at packages/package-stubber/main.js:159
I20140621-14:08:23.790(-7)?     at Function.forEach (/Users/sam/.meteorite/packages/package-stubber/alanning/meteor-package-stubber/bf8a11d648e266212867068b8f11280281129dbc/package-stubber/.npm/package/node_modules/lodash/dist/lodash.js:3297:15)
I20140621-14:08:23.790(-7)?     at Object._.extend.stubPackages (packages/package-stubber/main.js:154)
I20140621-14:08:23.790(-7)?     at Object.rerunTests (packages/jasmine-unit/main.js:121)
I20140621-14:08:23.790(-7)?     at observeChangesCallbacks.added (packages/minimongo/observe.js:153)
I20140621-14:08:23.790(-7)?     at self.applyChange.added (packages/minimongo/observe.js:53)
I20140621-14:08:23.790(-7)?     at packages/mongo-livedata/observe_multiplex.js:163
I20140621-14:08:23.796(-7)?     at Array.forEach (native)
```

this is because, global['Layout'] is undefined. I thought this is perhaps due to client side packages are not loaded so I added:

```
      if (global[name]) {
        DEBUG && console.log('[PackageStubber] stubbing', name)
        // `stubs` object will have one field of type `string` for each global
        // object that is being stubbed (ie. each package export)
        stubs[name] = PackageStubber.generateStubJsCode(global[name])
      } else {
        DEBUG && console.log('[PackageStubber] could not stub undefined', name)
      }
```

And the output was:

```
I20140621-14:10:19.617(-7)? [PackageStubber] could not stub undefined Layout
I20140621-14:10:19.617(-7)? [PackageStubber] stubbing RouteController
I20140621-14:10:19.651(-7)? Exception in queued task: Error: RouteController requires a router
I20140621-14:10:19.651(-7)?     at RouteController (packages/iron-router/lib/route_controller.js:37)
I20140621-14:10:19.651(-7)?     at Utils.extend.constructor (packages/iron-router/lib/server/route_controller.js:3)
I20140621-14:10:19.651(-7)?     at Object._.extend.generateStubJsCode (packages/package-stubber/main.js:369)
I20140621-14:10:19.651(-7)?     at packages/package-stubber/main.js:159
I20140621-14:10:19.652(-7)?     at Function.forEach (/Users/sam/.meteorite/packages/package-stubber/alanning/meteor-package-stubber/bf8a11d648e266212867068b8f11280281129dbc/package-stubber/.npm/package/node_modules/lodash/dist/lodash.js:3297:15)
I20140621-14:10:19.652(-7)?     at Object._.extend.stubPackages (packages/package-stubber/main.js:154)
I20140621-14:10:19.652(-7)?     at Object.rerunTests (packages/jasmine-unit/main.js:121)
I20140621-14:10:19.652(-7)?     at observeChangesCallbacks.added (packages/minimongo/observe.js:153)
I20140621-14:10:19.652(-7)?     at self.applyChange.added (packages/minimongo/observe.js:53)
I20140621-14:10:19.652(-7)?     at packages/mongo-livedata/observe_multiplex.js:163
```

I've not looked into the code deeply enough to figure it out so some light here would be awesome

**Comment by alanning** [comment]
I think that you are probably right re: Layout.  It's probably due to 'client' vs 'server' exports.  Currently package-stubber just treats everything the same so your fix to report "undefined" is a great idea.

The error re: RouteController is more worrisome.  I suspect that RouterController is a function (class) which requires a parameter in the constructor or else it will throw that error.  When package-stubber encounters a top-level export which is a function, it will execute it so that it can stub the resulting object.  This behavior was modeled after the `moment` library where you have things like:

`moment().format("MMM dd, YYYY")`

In the case where the top-level function requires a parameter, I'm not sure what we can do.  Maybe just wrap a try...catch around it and use a generic empty function as the stub in those cases?  That wouldn't be ideal though as there would probably be cases where the program expects a real object result and will call further functions on it.

I think this is just a limitation of auto-stubbing in general; I believe this would fail in the case of using the Jest library as well since there's no way to tell in Javascript when a function constructor requires a parameter.

So in general, I'd say we change the auto-stubber to catch those cases where function constructors fail when no parameters are supplied and also use special-case stubs where it makes sense.

For libraries like `iron-router` which are in heavy use I think it makes sense to have them special-cased, meaning:
1. package-stubber specifically ignores those libraries
2. an iron-router-specific stub is used instead

**Comment by SamHatoum** [comment]
I like the option of trying anyway to provide an empty method. If this allows the app to load statically, that's awesome. If not, the user will have had to do the work anyway. 

We could have some reasonable messaging that lets users know what they need to stub. Something like: "PackageStubber was unable to stub RouterController in the iron-router package. You'll need to manually stub this yourself if the app doesn't load".

And I agree that we should have the popular packages already stubbed.

**Comment by alanning** [comment]
OK, please test with this package-stubber branch:
https://github.com/alanning/meteor-package-stubber/tree/graceful-error-handling

It should correctly stub any type of target and also display useful error messages in cases where stubbing fails.

**Comment by gorru** [comment]
Hello, I'm trying to do some unit test with velocity patched with the change you suggest.
I'm getting this log:

```
I20140623-17:23:03.219(2)? [PackageStubber] Calling exported function 'RouteController' with no parameters produced an error. 'RouteController' has been stubbed with an empty function but if you receive errors due to missing fields in this package, you will need to supply your own custom stub. The original error was:  RouteController requires a router
I20140623-17:23:03.219(2)? [PackageStubber] Calling exported function 'Route' with no parameters produced an error. 'Route' has been stubbed with an empty function but if you receive errors due to missing fields in this package, you will need to supply your own custom stub. The original error was:  
I20140623-17:23:03.219(2)? [PackageStubber] Error generating stub for exported object 'Router'. 'Router' has been stubbed with an empty object but if you receive errors due to missing fields in this package, you will need to supply your own custom stub. The original error follows:
I20140623-17:23:03.220(2)?  Maximum call stack size exceeded
```

After that no test results is shown.

My smart.json contains

```
    "iron-router": {},
    "accounts-ui-bootstrap-dropdown": {},
    "spin": {},
    "jquery-ui": {},
    "jquery-ui-bootstrap": {},
    "publish-with-relations": {},
    "munit": {},
    "velocity": {},
    "jasmine-unit": {},
    "mocha-web": {}
```

Can you help me?
Thank's

**Comment by SamHatoum** [comment]
iron-router's RouteController and Route can't be stubbed automatically so we'll be creating a stub for it soon as it's a commonly used package.

**Comment by gorru** [comment]
Thank you.
There is some work around that I can try in order to bypass these errors and run my tests?

**Comment by alanning** [comment]
Hi @gorru,

You can stub the RouteController and Route yourself.  Add the stub code to a file in your tests folder named something like, "iron-router-stub.js".  The "-stub.js" part of the name will cause jasmine-unit to recognize its a stub and should be loaded in the proper order.

**Comment by alanning** [comment]
OK, this is fixed in `jasmine-unit` v0.1.11.  

It uses `package-stubber` v0.0.5 which supports custom stubs for popular packages and includes one for `iron-router`.  

Custom stubs will only be loaded for packages that the app actually uses.
